flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/LegacyStatefulJobSavepointMigrationITCase.java

```diff
@@ -0,0 +1,663 @@
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator;
+import org.apache.flink.streaming.api.operators.InternalTimer;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
+import org.apache.flink.streaming.api.operators.Triggerable;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.util.migration.MigrationVersion;
+import org.apache.flink.util.Collector;
+
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Migration ITCases for a stateful job. The tests are parameterized to cover
+ * migrating for multiple previous Flink versions, as well as for different state backends.
+ */
+@RunWith(Parameterized.class)
+public class LegacyStatefulJobSavepointMigrationITCase extends SavepointMigrationTestBase {
+
+	private static final int NUM_SOURCE_ELEMENTS = 4;
+
+	@Parameterized.Parameters(name = "Migrate Savepoint / Backend: {0}")
+	public static Collection<Tuple2<MigrationVersion, String>> parameters () {
+		return Arrays.asList(
+			Tuple2.of(MigrationVersion.v1_2, StateBackendLoader.MEMORY_STATE_BACKEND_NAME),
+			Tuple2.of(MigrationVersion.v1_2, StateBackendLoader.ROCKSDB_STATE_BACKEND_NAME),
+			Tuple2.of(MigrationVersion.v1_3, StateBackendLoader.MEMORY_STATE_BACKEND_NAME),
+			Tuple2.of(MigrationVersion.v1_3, StateBackendLoader.ROCKSDB_STATE_BACKEND_NAME),
+			Tuple2.of(MigrationVersion.v1_4, StateBackendLoader.MEMORY_STATE_BACKEND_NAME),
+			Tuple2.of(MigrationVersion.v1_4, StateBackendLoader.ROCKSDB_STATE_BACKEND_NAME));
+	}
+
+	/**
+	 * TODO to generate savepoints for a specific Flink version / backend type,
+	 * TODO change these values accordingly, e.g. to generate for 1.3 with RocksDB,
+	 * TODO set as (MigrationVersion.v1_3, StateBackendLoader.ROCKSDB_STATE_BACKEND_NAME)
+	 */
+	private final MigrationVersion flinkGenerateSavepointVersion = MigrationVersion.v1_4;
+	private final String flinkGenerateSavepointBackendType = StateBackendLoader.ROCKSDB_STATE_BACKEND_NAME;
+
+	private final MigrationVersion testMigrateVersion;
+	private final String testStateBackend;
+
+	public LegacyStatefulJobSavepointMigrationITCase(Tuple2<MigrationVersion, String> testMigrateVersionAndBackend) {
+		this.testMigrateVersion = testMigrateVersionAndBackend.f0;
+		this.testStateBackend = testMigrateVersionAndBackend.f1;
+	}
+
+	/**
+	 * Manually run this to write binary snapshot data.
+	 */
+	@Test
+	@Ignore
+	public void writeSavepoint() throws Exception {
```

Comment:
We could unify these methods.

Hide all function constructors behind a factory that accepts an enum `ExecutionMode(CREATE/RESTORE)` and add a switch at the end for deciding whether to call `executeAndSavepoint` and `restoreAndExecute`.

This would have the benefit that one test cannot be modified without becoming incompatible with the other.